test/e2e/specs/interactivity/instant-search.spec.ts

```diff
@@ -0,0 +1,650 @@
+			await expect( page ).not.toHaveURL(
+				new RegExp( `query-${ queryId }-page=2` )
+			);
+
+			// The url should now contain `?paged=1` because we're on the first page
+			// We cannot remove the `paged` param completely because the pathname
+			// might contain the `/page/2` suffix so we need to set `paged` to `1` to
+			// override it.
+			await expect( page ).toHaveURL(
+				new RegExp( `query-${ queryId }-page=1` )
+			);
+		} );
+
+		test( 'should show no-results block when search has no matches', async ( {
+			page,
+		} ) => {
+			await page
+				.locator( 'input[type="search"]' )
+				.fill( 'NonexistentContent' );
+			await page.waitForResponse( ( response ) =>
+				response
+					.url()
+					.includes(
+						`instant-search-${ queryId }=NonexistentContent`
+					)
+			);
+
+			// Verify no-results block is shown
+			await expect( page.getByText( 'No results found.' ) ).toBeVisible();
+		} );
+
+		test( 'should update pagination numbers based on search results', async ( {
+			page,
+		} ) => {
+			// Initially should show pagination numbers for 3 pages
+			await expect(
+				page.locator( '.wp-block-query-pagination-numbers' )
+			).toBeVisible();
+			await expect(
+				page.getByRole( 'link', { name: '2' } )
+			).toBeVisible();
+			await expect(
+				page.getByRole( 'link', { name: '3' } )
+			).toBeVisible();
+
+			// Search for unique post
+			await page.locator( 'input[type="search"]' ).fill( 'Unique' );
+			await page.waitForResponse( ( response ) =>
+				response.url().includes( `instant-search-${ queryId }=Unique` )
+			);
+
+			// Pagination numbers should not be visible with single result
+			await expect(
+				page.locator( '.wp-block-query-pagination-numbers' )
+			).toBeHidden();
+		} );
+
+		test( 'should handle pre-defined search from query attributes', async ( {
+			requestUtils,
+			page,
+		} ) => {
+			// Create page with custom query that includes a search parameter
+			const { id } = await requestUtils.createPage( {
+				status: 'publish',
+				title: 'Query with Search',
+				content: `
+<!-- wp:query {"enhancedPagination":true,"queryId":${ queryId },"query":{"inherit":false,"perPage":2,"order":"desc","orderBy":"date","offset":0,"search":"Unique"}} -->
+    <div class="wp-block-query" data-testid="query-with-search">
+        <!-- wp:search {"label":"","buttonText":"Search"} /-->
+        <!-- wp:post-template -->
+            <!-- wp:post-title {"level":3} /-->
+            <!-- wp:post-excerpt /-->
+        <!-- /wp:post-template -->
+        <!-- wp:query-pagination -->
+            <!-- wp:query-pagination-previous /-->
+            <!-- wp:query-pagination-numbers /-->
+            <!-- wp:query-pagination-next /-->
+        <!-- /wp:query-pagination -->
+        <!-- wp:query-no-results -->
+            <!-- wp:paragraph -->
+            <p>No results found.</p>
+            <!-- /wp:paragraph -->
+        <!-- /wp:query-no-results -->
+    </div>
+<!-- /wp:query -->`,
```

Comment:
Nit pick but IMO using strings to set blocks is like using a snapshot which should not be the case in this context. We do the same in the zoom out context setting block content from strings which can change and invalidate the test for no reason.

Reply:
I'm not sure I understand. I think they are not like snapshots because if I change some whitespace in this string, it will still result in the same block structure in the editor. And it's that block structure that we're testing. Am I missing something?🙂